src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp

```diff
@@ -7250,6 +7251,349 @@ class StubGenerator: public StubCodeGenerator {
+
+      __ poly1305_field_multiply(u1, u0_lo, R, RR2, regs);
+      // u1_lo = R**3
+
+      // RR2 = 5 * (R[2] << 26)
+      __ copy_3_regs(R, u1_lo);
+      __ lsl(RR2, R[2], 26);
+      __ add(RR2, RR2, RR2, __ LSL, 2);
+
+      __ poly1305_field_multiply(u1, R, R, RR2, regs);
+      // u1_lo = R**6
+
+      __ copy_3_regs(R, u1_lo);
+      __ lsl(RR2, R[2], 26);
+      __ add(RR2, RR2, RR2, __ LSL, 2);
+    }
+
+    // Load the initial state
+    __ pack_26(u0[0]._lo, u0[1]._lo, u0[2]._lo, acc_start);
+
+    // u0 contains the initial state. Clear the others.
+    for (int i = 0; i < 3; i++) {
+      __ mov(u0[i]._hi, 0);
+      __ mov(u1[i]._lo, 0); __ mov(u1[i]._hi, 0);
+    }
+
+    VectorRegs v_u0(vregs, 5);
+    VectorRegs v_s0(vregs, 3);
+    VectorRegs v_u1(vregs, 5);
+    VectorRegs v_s1(vregs, 3);
+
+    const FloatRegister zero = *vregs++;
+
+    __ movi(zero, __ T16B, 0);
+
+    // rr_v = r_v * 5
+    VectorRegs r_v(vregs, 2);
+    VectorRegs rr_v(vregs, 2);
+    __ copy_3_regs_to_5_elements(r_v, R[0], R[1], R[2]);
+    {
+      FloatRegister vtmp = *vregs;
+      __ shl(vtmp, __ T4S, r_v[0], 2);
+      __ addv(rr_v[0], __ T4S, r_v[0], vtmp);
+      __ shl(vtmp, __ T4S, r_v[1], 2);
+      __ addv(rr_v[1], __ T4S, r_v[1], vtmp);
+    }
+
+    for (int i = 0; i < 5; i++) {
+      __ movi(v_u0[i], __ T16B, 0);
+      __ movi(v_u1[i], __ T16B, 0);
+    }
+
+    {
+      Label DONE, LOOP;
+
+      __ subsw(rscratch1, length, POLY1305_BLOCK_LENGTH * BLOCKS_PER_ITERATION * 2);
+      __ br(Assembler::LT, DONE);
+
+      __ align(OptoLoopAlignment);
+      __ bind(LOOP);
+      {
+        constexpr int COLS = 4;
+        AsmGenerator gen[COLS];
+
+        __ poly1305_step(gen[0], S0, u0, input_start);
+        __ poly1305_field_multiply(gen[0], u0, S0, R, RR2, regs);
+
+        __ poly1305_step(gen[1], S1, u1, input_start);
+        __ poly1305_field_multiply(gen[1], u1, S1, R, RR2, regs);
+
+        __ poly1305_step_vec(gen[2], v_s0, v_u0, zero, input_start);
+        __ poly1305_field_multiply(gen[2], v_u0, v_s0, r_v, rr_v, zero,
+                                   vregs.remaining());
+
+        __ poly1305_step_vec(gen[3], v_s1, v_u1, zero, input_start);
+        __ poly1305_field_multiply(gen[3], v_u1, v_s1, r_v, rr_v, zero,
+                                   vregs.remaining());
+
+        AsmGenerator::Iterator it[COLS];
+        int len[COLS];
+
+        int l_max = INT_MIN;
+        for (int col = 0; col < COLS; col++) {
+          it[col] = gen[col].iterator();
+          len[col] = gen[col].length();
+          l_max = MAX2(l_max, len[col]);
+        }
+
+        int err[COLS];
+        for (int col = 0; col < COLS; col++) {
+          err[col] = 0;
+        }
+
+        for (int i = 0; i < l_max; i++) {
+          for (int col = 0; col < COLS; col++) {
+            err[col] -= len[col];
+            if (err[col] < 0) {
+              err[col] += l_max;
+              (it[col]++)();
+            }
+          }
+        }
+
```

Comment:
I forgot to add some feedback on this assert and the preceding loop. I think it would be good to explain what is going on here

    // The column countdowns and updates in the above loop
    // are designed to schedule execution of the instruction
    // generation closures in each of the generator queues
    // so as to interleave the resulting parallel instruction
    // streams as evenly as possible. It should be self-evident
    // that the following assertion must hold (and only hold)
    // once all the queues have been exhausted.